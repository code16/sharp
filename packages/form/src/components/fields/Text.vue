<template>
    <input 
        class="SharpText"
        :type="inputType" 
        :value="value"
        :placeholder="placeholder"
        :disabled="readOnly"
        v-maxlength="maxLength"
        @input="handleInput"
        ref="input"
    >
</template>

<script>
    import { Focusable } from 'sharp/mixins';
    import { maxlength } from 'sharp/directives';

    export default {
        name:'SharpText',

        mixins: [Focusable],

        props: {
            value: [String, Number],

            placeholder: String,
            readOnly: Boolean,

            maxLength:Number,

            inputType:  {
                type:String,
                default:'text'
            },
        },
        data() {
            return {}
        },
        methods: {
            handleInput(e) {
                this.$emit('input',e.target.value);
            }
        },

        mounted() {
            this.setFocusable(this.$refs.input);
        },

        directives: {
            maxlength
        }
    }
</script>