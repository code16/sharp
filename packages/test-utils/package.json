{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}