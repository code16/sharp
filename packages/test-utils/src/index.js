export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}