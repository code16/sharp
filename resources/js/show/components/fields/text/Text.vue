<script setup lang="ts">
    import { __ } from "@/utils/i18n";
    import { ShowTextFieldData } from "@/types";
    import { computed, provide, ref } from "vue";
    import ShowFieldLayout from "../../ShowFieldLayout.vue";
    import TextRenderer from "./TextRenderer.vue";
    import clip from "text-clipper";
    import { ShowFieldProps } from "@/show/types";
    import { ContentEmbedManager } from "@/content/ContentEmbedManager";
    import { useParentShow } from "@/show/useParentShow";
    import { ContentUploadManager } from "@/content/ContentUploadManager";
    import { Button } from "@/components/ui/button";

    const props = defineProps<ShowFieldProps<ShowTextFieldData>>();

    const expanded = ref(false);
    const show = useParentShow();

    const embedManager = new ContentEmbedManager(show, props.field.embeds, props.value?.embeds);
    const uploadManager = new ContentUploadManager(show, props.value?.uploads);

    provide('embedManager', embedManager);
    provide('uploadManager', uploadManager);

    const localizedValue = computed<string | null>(() => {
        return props.field.localized
            ? props.value.text?.[props.locale]
            : props.value?.text as string;
    });

    function stripTags(html) {
        const el = document.createElement('div');
        el.innerHTML = html;
        return el.textContent;
    }

    function truncateToWords(text, count) {
        const matches = [...text.matchAll(/\S+\s*/g)];
        return matches.length > count
            ? matches.slice(0, count).map(match => match[0]).join('')
            : text;
    }

    const collapsedContent = computed(() => {
        const { field } = props;
        const value = localizedValue.value?.trim();
        if (!field.collapseToWordCount || !value) {
            return null;
        }
        const text = stripTags(value).trim();
        const content = field.html ? value.trim() : text;
        const truncated = truncateToWords(text, field.collapseToWordCount);
        return truncated.length < text.length
            ? clip(content, truncated.length + 2, field.html ? { html: true } : {})
            : null;
    });

    const currentContent = computed(() => {
        if(!localizedValue.value) {
            return null;
        }
        if(collapsedContent.value && !expanded.value) {
            return collapsedContent.value;
        }
        if(!props.field.html) {
            return stripTags(localizedValue.value).trim();
        }
        return localizedValue.value;
    });
</script>

<template>
    <ShowFieldLayout v-bind="props">
        <template v-if="currentContent && field.html">
            <TextRenderer
                class="content content-sm text-sm [:where(&)_:where(h1,h2,h3)]:text-foreground/75"
                :content="currentContent"
                :field="field"
            />
        </template>
        <template v-else>
            <div class="text-sm">
                {{ currentContent }}
            </div>
        </template>

        <template v-if="collapsedContent">
            <Button class="mt-2 px-0" variant="link" @click.prevent="expanded = !expanded">
                <template v-if="expanded">
                    - {{ __('sharp::show.text.show_less') }}
                </template>
                <template v-else>
                    + {{ __('sharp::show.text.show_more') }}
                </template>
            </Button>
        </template>
    </ShowFieldLayout>
</template>
