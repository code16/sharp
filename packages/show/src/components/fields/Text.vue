<template>
    <FieldLayout class="ShowTextField" :label="label">
        <div class="ShowTextField__content" v-html="currentContent" ref="content"></div>
        <template v-if="hasCollapsed">
            <div class="mt-2">
                <a href="#" class="ShowTextField__more" @click.prevent="handleToggleClicked">
                    <template v-if="expanded">
                        - {{ l('show.text.show_less') }}
                    </template>
                    <template v-else>
                        + {{ l('show.text.show_more') }}
                    </template>
                </a>
            </div>
        </template>
    </FieldLayout>
</template>

<script>
    import { Localization } from 'sharp/mixins';
    import clip from 'text-clipper';
    import { syncVisibility } from "../../util/fields/visiblity";
    import { truncateToWords, stripTags } from "../../util/fields/text";
    import FieldLayout from "../FieldLayout";

    export default {
        mixins: [Localization],
        components: {
            FieldLayout,
        },
        props: {
            value: String,
            collapseToWordCount: Number,
            label: String,
            emptyVisible: Boolean,
        },
        data() {
            return {
                expanded: false,
            }
        },
        computed: {
            currentContent() {
                if(this.hasCollapsed && !this.expanded) {
                    return this.collapsedContent;
                }
                return this.value;
            },
            hasCollapsed() {
                return !!this.collapsedContent;
            },
            collapsedContent() {
                if(!this.collapseToWordCount || !this.value) {
                    return null;
                }
                const html = this.value.trim();
                const text = stripTags(html);
                const truncated = truncateToWords(text, this.collapseToWordCount);
                return truncated.length < text.length
                    ? clip(html, truncated.length + 2, { html: true })
                    : null;
            },
            isVisible() {
                return !!this.value || this.emptyVisible;
            },
        },
        methods: {
            handleToggleClicked() {
                this.expanded = !this.expanded;
            },
        },
        created() {
            syncVisibility(this, () => this.isVisible);
        },
    }
</script>