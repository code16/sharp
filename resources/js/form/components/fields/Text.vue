<script setup lang="ts">
    import { FormTextFieldData } from "@/types";
    import { ref } from "vue";
    import { normalizeText } from "../../util/text";
    import { validateTextField } from "../../util/validation";
    import { FormFieldEmits, FormFieldProps } from "@/form/types";
    import FormFieldLayout from "@/form/components/FormFieldLayout.vue";
    import { Input } from "@/components/ui/input";
    import { Button } from "@/components/ui/button";
    import { Eye, EyeOff } from 'lucide-vue-next';

    const props = defineProps<FormFieldProps<FormTextFieldData>>();
    const emit = defineEmits<FormFieldEmits<FormTextFieldData>>();

    const input = ref();
    const passwordVisible = ref(false);

    function onInput(inputValue: string) {
        const value = normalizeText(inputValue);
        const error = validateTextField(value, {
            maxlength: props.field.maxLength,
        });
        if(props.field.localized) {
            emit('input',
                typeof props.value === 'object'
                    ? { ...props.value, [props.locale]: value }
                    : { [props.locale]: value },
                { error }
            );
        } else {
            emit('input', value, { error });
        }
    }

    defineExpose({
        focus: () => input.value.$el.focus(),
    });
</script>

<template>
    <FormFieldLayout v-bind="props" v-slot="{ id, ariaDescribedBy }">
        <div class="relative">
            <Input
                :id="id"
                :class="field.inputType === 'password' ? 'pr-10' : ''"
                :model-value="field.localized && typeof value === 'object' ? value?.[locale] : (value as string)"
                :placeholder="field.placeholder"
                :disabled="field.readOnly"
                :aria-describedby="ariaDescribedBy"
                :type="passwordVisible ? 'text' : field.inputType"
                @update:model-value="onInput"
                ref="input"
            />
            <template v-if="field.inputType === 'password'">
                <Button class="absolute size-9.5 right-px top-px rounded-[calc(var(--radius)-3px)]" size="icon" variant="ghost" @click="passwordVisible = !passwordVisible">
                    <template v-if="passwordVisible">
                        <EyeOff />
                    </template>
                    <template v-else>
                        <Eye />
                    </template>
                </Button>
            </template>
        </div>
    </FormFieldLayout>
</template>
