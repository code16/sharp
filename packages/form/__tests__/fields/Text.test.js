import Vue from 'vue';
import Text from '../../src/components/fields/Text.vue';

import { MockInjections } from '@sharp/test-utils';

describe('text-field', () => {
    Vue.component('sharp-text', Text);

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-text :input-type="inputType" 
                            placeholder="Entrez du texte" 
                            value="AAA" read-only 
                            @input="inputEmitted($event)">
                </sharp-text>
            </div>
        `
    });

    test('can mount Text field', async () => {
        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('emit event on input and correct value', async () => {
        let inputEmitted = jest.fn();

        await createVm({
            methods: {
                inputEmitted
            }
        });
        let input = document.querySelector('input');
        input.dispatchEvent(new Event('input', { bubbles:true }));
        expect(inputEmitted).toHaveBeenCalledTimes(1);
        expect(inputEmitted).toHaveBeenCalledWith('AAA');
    });

    test('take input type in account', async () => {
        await createVm({
            propsData: {
                inputType: 'password'
            }
        });
        let input = document.querySelector('input');
        expect(input.type).toBe('password');
    });

});

async function createVm(customOptions={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [MockInjections, customOptions],

        props:['inputType']
    });

    await Vue.nextTick();

    return vm.$children[0];
}