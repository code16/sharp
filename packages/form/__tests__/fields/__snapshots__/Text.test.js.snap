// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text-field can mount Text field 1`] = `
<div id="app">
	<input type="text" placeholder="Entrez du texte" disabled="disabled" class="SharpText">
</div>
`;
